File: python/cudf_streaming/cudf_streaming/tests/test_partition.py
# SPDX-FileCopyrightText: Copyright (c) 2025-2026, NVIDIA CORPORATION & AFFILIATES.
# SPDX-License-Identifier: Apache-2.0

from __future__ import annotations

from typing import TYPE_CHECKING

import pylibcudf as plc
import pytest

from cudf_streaming.streaming.partition import (
    partition_and_pack,
    unpack_and_concat,
)
from cudf_streaming.streaming.table_chunk import TableChunk
from rapidsmpf.streaming.core.actor import run_actor_network
from rapidsmpf.streaming.core.leaf_actor import (
    pull_from_channel,
    push_to_channel,
)
from rapidsmpf.streaming.core.message import Message
from rapidsmpf.testing import assert_eq

if TYPE_CHECKING:
    from rapidsmpf.streaming.chunks.partition import PartitionMapChunk
    from rapidsmpf.streaming.core.channel import Channel
    from rapidsmpf.streaming.core.context import Context
    from rmm.pylibrmm.stream import Stream


@pytest.mark.parametrize("num_partitions", [1, 2, 3, 10])
def test_partition_and_pack_unpack(
    context: Context, stream: Stream, num_partitions: int
) -> None:
    expects = [
        plc.Table(
            [
                plc.Column.from_iterable_of_py(
                    [1, 2, 3], plc.DataType(plc.TypeId.INT64)
                ),
                plc.Column.from_iterable_of_py(
                    [2, 1, 1], plc.DataType(plc.TypeId.INT64)
                ),
            ]
        ),
        plc.Table(
            [
                plc.Column.from_iterable_of_py(
                    [], plc.DataType(plc.TypeId.INT64)
                ),
                plc.Column.from_iterable_of_py(
                    [], plc.DataType(plc.TypeId.INT64)
                ),
            ]
        ),
    ]
    table_chunks = [
        Message(
            seq,
            TableChunk.from_pylibcudf_table(
                expect, stream, exclusive_view=False, br=context.br()
            ),
        )
        for seq, expect in enumerate(expects)
    ]
    ch1: Channel[TableChunk] = context.create_channel()
    actor1 = push_to_channel(context, ch_out=ch1, messages=table_chunks)

    ch2: Channel[PartitionMapChunk] = context.create_channel()
    actor2 = partition_and_pack(
        context,
        ch_in=ch1,
        ch_out=ch2,
        columns_to_hash=(1,),
        num_partitions=num_partitions,
    )

    ch3: Channel[TableChunk] = context.create_channel()
    actor3 = unpack_and_concat(
        context,
        ch_in=ch2,
        ch_out=ch3,
    )

    actor4, output = pull_from_channel(context, ch_in=ch3)
    run_actor_network(context, actors=(actor1, actor2, actor3, actor4))

    results = output.release()
    for seq, (result, expect) in enumerate(zip(results, expects, strict=True)):
        assert result.sequence_number == seq
        tbl = TableChunk.from_message(result, br=context.br())
        assert_eq(tbl.table_view(), expect, sort_rows=0)

File: cpp/libcudf_streaming/tests/streaming/test_partition.cpp
/**
 * SPDX-FileCopyrightText: Copyright (c) 2025-2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * reserved. SPDX-License-Identifier: Apache-2.0
 */

#include "../utils.hpp"
#include "base_streaming_fixture.hpp"

#include <cudf_test/cudf_gtest.hpp>
#include <cudf_test/table_utilities.hpp>

#include <cudf_streaming/partition.hpp>
#include <cudf_streaming/table_chunk.hpp>

#include <gmock/gmock.h>
#include <rapidsmpf/communicator/single.hpp>
#include <rapidsmpf/memory/buffer.hpp>
#include <rapidsmpf/shuffler/chunk.hpp>
#include <rapidsmpf/streaming/chunks/partition.hpp>
#include <rapidsmpf/streaming/core/context.hpp>
#include <rapidsmpf/streaming/core/leaf_actor.hpp>

using namespace cudf_streaming;

using StreamingPartition = BaseStreamingFixture;

TEST_F(StreamingPartition, PackUnpackRoundTrip)
{
  int const num_partitions              = 5;
  int const num_rows                    = 100;
  int const num_chunks                  = 10;
  std::int64_t const seed               = 42;
  constexpr cudf::hash_id hash_function = cudf::hash_id::HASH_MURMUR3;

  std::vector<cudf::table> expects;
  for (int i = 0; i < num_chunks; ++i) {
    expects.push_back(random_table_with_index(seed + i, num_rows, 0, 10));
  }

  std::vector<rapidsmpf::streaming::Message> inputs;
  for (int i = 0; i < num_chunks; ++i) {
    inputs.emplace_back(to_message(
      i,
      std::make_unique<table_chunk>(
        std::make_unique<cudf::table>(expects[i], stream, ctx->br()->device_mr()), stream)));
  }

  // Create and run the streaming pipeline.
  std::vector<rapidsmpf::streaming::Message> outputs;
  {
    std::vector<rapidsmpf::streaming::Actor> actors;
    auto ch1 = ctx->create_channel();
    actors.push_back(rapidsmpf::streaming::actor::push_to_channel(ctx, ch1, std::move(inputs)));

    auto ch2 = ctx->create_channel();
    actors.push_back(cudf_streaming::actor::partition_and_pack(
      ctx, ch1, ch2, {1}, num_partitions, hash_function, seed));

    auto ch3 = ctx->create_channel();
    actors.push_back(cudf_streaming::actor::unpack_and_concat(ctx, ch2, ch3));

    actors.push_back(rapidsmpf::streaming::actor::pull_from_channel(ctx, ch3, outputs));

    rapidsmpf::streaming::run_actor_network(std::move(actors));
  }

  EXPECT_EQ(expects.size(), outputs.size());
  for (std::size_t i = 0; i < expects.size(); ++i) {
    EXPECT_EQ(outputs[i].sequence_number(), i);
    auto output = outputs[i].release<table_chunk>();
    CUDF_TEST_EXPECT_TABLES_EQUIVALENT(sort_table(output.table_view()),
                                       sort_table(expects[i].view()));
  }
}
